src/averageAges.js

```diff
@@ -55,10 +73,36 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const mothers = people.filter(mother => people.some(person =>
+    person.mother === mother.name));
+
+  let kids = people.filter(mummy => mummy.mother != null);
+
+  if (onlyWithSon) {
+    kids = people.filter(son => son.mother != null && son.sex === 'm');
+  }
+
+  let ageSum = 0;
+  let numberOfPairs = 0;
+
+  kids.forEach(kid => {
+    const mother = mothers.find(mummy => mummy.name === kid.mother);
```

Comment:
The usage of forEach() is discouraged as the task explicitly mentioned to avoid using loop and forEach. You can replace the forEach with a reduce function.